packages/hub/src/components/NavigationBlocker/context.ts

```diff
@@ -0,0 +1,69 @@
+import { createContext, Reducer } from "react";
+
+type State = {
+  // Blockers are functions; this way we can check dynamic parameters such as
+  // form values to decide whether navigation should be blocked.
+  blockers: Record<string, () => boolean>;
+  // Link was intercepted, need to show a modal.
+  interceptedLink: string | undefined;
+};
+
+type Action =
+  | {
+      type: "addBlocker";
+      payload: {
+        key: string;
+        blocker: () => boolean;
+      };
+    }
+  | {
+      type: "removeBlocker";
+      payload: {
+        key: string;
+      };
+    }
+  | {
+      type: "intercept";
+      payload: {
+        link: string;
+      };
+    }
+  | {
+      type: "clearInterceptedLink";
+    };
+
+export const navigationBlockerReducer: Reducer<State, Action> = (
```

Comment:
I found it surprising that you chose to use a formal Reducer here, but I generally like this pattern. It helps make things explicit.

Reply:
The interesting reason for this is that it's easier to have a single `dispatch` function in the provider, instead of multiple callbacks.

With callbacks I'd have to write out the type of each callback in the context type, and type definitions become more scattered over the codebase (e.g. you use `const [foo, setFoo] = useState<FooShape>()` in your wrapper component, and then you also have `{ setFoo: (v: FooShape) => void }` in your context type).